lib/graphql/pagination/relation_connection.rb

```diff
@@ -156,32 +162,39 @@ def after_offset
       def limited_nodes
         @limited_nodes ||= begin
           paginated_nodes = sliced_nodes
-          previous_limit = relation_limit(paginated_nodes)
+          @relation_limit ||= relation_limit(paginated_nodes)
 
-          if first && (previous_limit.nil? || previous_limit > first)
+          if first && (@relation_limit.nil? || @relation_limit > first)
             # `first` would create a stricter limit that the one already applied, so add it
-            paginated_nodes = set_limit(paginated_nodes, first)
+            @relation_limit = first
+            # paginated_nodes = set_limit(paginated_nodes, first)
           end
 
           if last
-            if (lv = relation_limit(paginated_nodes))
-              if last <= lv
+            if @relation_limit
+              if last <= @relation_limit
                 # `last` is a smaller slice than the current limit, so apply it
-                offset = (relation_offset(paginated_nodes) || 0) + (lv - last)
+                offset = (relation_offset(paginated_nodes) || 0) + (@relation_limit - last)
                 paginated_nodes = set_offset(paginated_nodes, offset)
-                paginated_nodes = set_limit(paginated_nodes, last)
+                @relation_limit = last
+                # paginated_nodes = set_limit(paginated_nodes, last)
               end
             else
               # No limit, so get the last items
               sliced_nodes_count = relation_count(@sliced_nodes)
               offset = (relation_offset(paginated_nodes) || 0) + sliced_nodes_count - [last, sliced_nodes_count].min
               paginated_nodes = set_offset(paginated_nodes, offset)
-              paginated_nodes = set_limit(paginated_nodes, last)
+              @relation_limit = last
             end
           end
 
           @paged_nodes_offset = relation_offset(paginated_nodes)
-          paginated_nodes
+
+          if paginated_nodes.loaded?
+            paginated_nodes.take(@relation_limit)
+          else
+            paginated_nodes.limit(@relation_limit)
+          end
```

Comment:
This worked, but I think this if/else block can potentially just be replaced with only?

```
paginated_nodes.take(@relation_limit)
```

because `take` itself is also checking if the records are `loaded?`:

https://github.com/rails/rails/blob/a35a380c2cdc7fe032f927095545aee21a192f54/activerecord/lib/active_record/relation/finder_methods.rb#L524

I tried this out and it worked for both a preloaded association and one that wasn't preloaded. Although I'm guessing just calling `take` would break the `SequelDatasetConnection` and `MongoidRelationConnection` and another hook method would need to be added? But I guess that's already happening with `loaded?`.

Reply:
Yeah, definitely agreed that this can be cleaned up one way or another, as long as it works with the different backends :+1: :eyes: